docs/developer-guide/join-streams-and-tables.rst

```diff
@@ -0,0 +1,329 @@
+
+**Co-partitioned data**: Input data must be co-partitioned when joining. This ensures
+that input records with the same key, from both sides of the join, are delivered
+to the same stream task during processing. It’s your responsibility to ensure data
+co-partitioning when joining. For more info, see Partition Data to Enable Joins [TBD new topic].
+
+**KEY property**: If you set the KEY property in a KSQL query, ensure that both of the following
+conditions are true:
+
+* For every record, the contents of the message key of the Kafka message itself must be
+  the same as the contents of the column set in KEY.
+* The KEY property must be set to a column of type VARCHAR or STRING.
+
+TBD: code snippet
+
+For more information, see :ref:`ksql_key_requirements`.
+
+Join Capabilities
+*****************
+
+KSQL supports a large set of join operations for streams and tables, including
+INNER, LEFT OUTER, and FULL OUTER. Frequently, LEFT OUTER is shortened to LEFT JOIN,
+and FULL OUTER is shortened to OUTER JOIN.
+
+.. note:: RIGHT OUTER JOIN isn’t supported. Instead, swap the operands and use LEFT JOIN.
+
+The following table shows the supported combinations.
+
++---------------+--------------+-----------+------------+---------------+
+|               | Type         | INNER     | LEFT OUTER | FULL OUTER    |
++===============+==============+===========+============+===============+
+| Stream-Stream | Windowed     | Supported | Supported  | Supported     |                                         
++---------------+--------------+-----------+------------+---------------+
+| Table-Table   | Non-windowed | Supported | Supported  | Supported     |
++---------------+--------------+-----------+------------+---------------+
+| Stream-Table  | Non-windowed | Supported | Supported  | Not supported |
++---------------+--------------+-----------+------------+---------------+
+
+Stream-Stream Joins
+*******************
+
+KSQL supports INNER, LEFT OUTER, and FULL OUTER, joins between streams.
+
+All of these operations support out-of-order records.
+
+Joins between streams are always windowed joins. A new input record on one side
+produces a join output for each matching record on the other side, and there
+can be multiple such matching records within a join window.
+
+INNER joins cause data re-partitioning of a stream only if the stream was
+marked for re-partitioning If both streams are marked, both are re-partitioned.
+
+LEFT OUTER joins will contain leftRecord-NULL records in the result stream,
+which means that the join contains NULL values for fields selected from the
+right-hand stream where no match is made.
+
+FULL OUTER joins will contain leftRecord-NULL or NULL-rightRecord records in
+the result stream, which means that the join contains NULL values for fields
+coming from a stream where no match is made.
+
+Semantics of Stream-Stream Joins
+--------------------------------
+
+The semantics of the various stream-stream join variants are shown in the

```

Comment:
I assume the sections on stream-stream, stream-table, and table-table joins are lifted from the streams docs directly and don't need further review?

Reply:
Yes indeed, I just replaced "KStream" with "Stream" and "KTable" with "Table".